config/config-api/src/com/thoughtworks/go/domain/PipelineGroups.java

```diff
@@ -158,18 +158,17 @@ public void addError(String fieldName, String message) {
         configErrors.add(fieldName, message);
     }
 
-    public HashSet<MaterialConfig> getAllUniquePostCommitSchedulableMaterials() {
-        HashSet<MaterialConfig> materialConfigs = new HashSet<MaterialConfig>();
-        Set<Map> uniqueMaterials = new HashSet<Map>();
+    public Set<MaterialConfig> getAllUniquePostCommitSchedulableMaterials() {
+        Set<MaterialConfig> materialConfigs = new HashSet<MaterialConfig>();
+        Set<String> uniqueMaterials = new HashSet<String>();
         for (PipelineConfigs pipelineConfigs : this) {
             for (PipelineConfig pipelineConfig : pipelineConfigs) {
                 for (MaterialConfig materialConfig : pipelineConfig.materialConfigs()) {
-                    if (!uniqueMaterials.contains(materialConfig.getSqlCriteria()) && materialConfig instanceof ScmMaterialConfig) {
-                        if (materialConfig.isAutoUpdate()) {
-                            continue;
-                        }
+                    if ((materialConfig instanceof ScmMaterialConfig || materialConfig instanceof PluggableSCMMaterialConfig)
+                            && !materialConfig.isAutoUpdate()
+                            && !uniqueMaterials.contains(materialConfig.getFingerprint())) {
                         materialConfigs.add(materialConfig);
-                        uniqueMaterials.add(materialConfig.getSqlCriteria());
+                        uniqueMaterials.add(materialConfig.getFingerprint());
```

Comment:
Ok, see about the null. My worry about the fingerprint is performance, obviously. If you look, ```getAllUniquePostCommitSchedulableMaterials``` is not cached at all (even though it goes through every group, pipeline and material). When config changes, all these objects should get thrown away, so locally caching this should be possible, I suppose. Right?

Only case this can fail is that a plugin which was missing is brought back. I suppose that's a very corner case.